src/core/labeling/qgspallabeling.cpp

```diff
@@ -2373,119 +2373,159 @@ std::unique_ptr<QgsLabelFeature> QgsPalLayerSettings::registerFeatureWithDetails
     }
   }
 
-  if ( mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionX ) )
+  bool hasDataDefinedPosition = false;
   {
-    exprVal = mDataDefinedProperties.value( QgsPalLayerSettings::PositionX, context.expressionContext() );
-    if ( !exprVal.isNull() )
+    bool ddPosition = false;
+
+    if ( mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionX )
+         && mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionY )
+         && !mDataDefinedProperties.value( QgsPalLayerSettings::PositionX, context.expressionContext() ).isNull()
+         && !mDataDefinedProperties.value( QgsPalLayerSettings::PositionY, context.expressionContext() ).isNull() )
```

Comment:
We are double-evaluating these two properties -- can you store the value from the first evaluation to avoid this?